Removing unused code.

paulp opened the pull request:
Will add some details in email.

paulp committed:
Removed unused imports in swing.

paulp committed:
Remove unused imports in partest.

paulp committed:
Remove unused imports in continuations.

paulp committed:
Deal with possibly spurious warning in Macros.

paulp committed:
Remove unused imports in library.

paulp committed:
applyOrElse is a synthetic method.

paulp committed:
Commenting out unused members.

I want to get this commit into the history because
the tests pass here, which demonstrates that every commented
out method is not only unnecessary internally but has zero
test coverage. Since I know (based on the occasional source
code comment, or more often based on knowing something about
other source bases) that some of these can't be removed
without breaking other things, I want to at least record
a snapshot of the identities of all these unused and
untested methods.

This commit will be reverted; then there will be another
commit which removes the subset of these methods which I
believe to be removable. The remainder are in great need of
tests which exercise the interfaces upon which other
repositories depend.

paulp committed:
Revert "Commenting out unused members."

This reverts commit 951fc3a486.

paulp committed:
Removing ancient comments and pointless comments.

Translating <code></code> into backticks.

Removed the "@param tree ..." blocks which have been
taunting me for half a decade now.

Removed commented-out blocks of code which had been
sitting there for two years or more.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1630/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1630/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/925/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/925/

xeno-by commented on src/compiler/scala/reflect/reify/phases/Metalevels.scala:
```diff
@@ -101,7 +103,7 @@ trait Metalevels {
   val metalevels = new Transformer {
     var insideSplice = false
-    var inlineableBindings = scala.collection.mutable.Map[TermName, Tree]()
+    val inlineableBindings = mutable.Map[TermName, Tree]()
```
Could you explain this part please?

paulp replied:
Placing a mutable map in a var is always the wrong thing. Either "var" or "mutable" give you all the mutation you can eat.

xeno-by replied:
Oh didn't notice that part. I meant why refactor `scala.collection.mutable` into `mutable`?

paulp replied:
Super long paths distract from what is going on. There is no benefit to having "scala.collection" sprinkled around. Also, consistency.

viktorklang replied:
Totally agree

xeno-by commented on src/compiler/scala/tools/nsc/ast/Printers.scala:
```diff
@@ -200,91 +200,12 @@ trait Printers extends scala.reflect.internal.Printers { this: Global =>
     override def printTree(tree: Tree) { print(safe(tree)) }
   }
 
-  class TreeMatchTemplate {
```
I think this at least deserves being moved to the docs

paulp replied:
Since I wrote it I can't dispute that it is a national treasure. Moving it to docs is a failure though, then it just rots. I'll rescue it and use it in a test case.

jsuereth commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -62,30 +62,23 @@ class MutableSettings(val errorFn: String => Unit)
+      // but discard carefully, because an empty string is valid as an argument
+      // to an option, e.g. -cp "" .  So we discard them only when they appear
+      // where an option should be, not where an argument to an option should be.
+      case "" :: xs =>
```
Just a note that this area of the compiler lacks a good automated test.  We should manually run the ant tasks and command line tools to validate nothing wonky happens here, or write tests.

jsuereth commented:
Is there any chance we can split this into even smaller, more manageable chunks for review?   I'm not sure it's humanly possible for someone to look through this and validate it's ok.   

It's a very impressive and amazing piece of work, but I think it'd be far easier for us to digest and accept in piecemeal fashion.

Another good thing would be moving extraneous commits, like the argument parsing code, into a separate pull request.

Thanks for this amazing piece of work!

paulp committed:
Some files removed in their entirety.

And small associated changes.

paulp committed:
Members removed from the repl.

There were a whole lot of these.

paulp committed:
Members removed from scala.reflect.io.

Mostly hailing from a long-ago day when I imagined I was
writing a general purpose library. We dodged that bullet.

paulp committed:
Members removed from partest.

It has accreted its share through the bumpy years.

paulp committed:
Members removed in the backend.

paulp committed:
Members removed in scala.reflect.

Not a bad showing for a newcomer. Of course most of this
code predates scala.reflect by a lot.

paulp committed:
Removed code from the typechecker.

Removing code from this neighborhood is more difficult than
elsewhere, making it all the more important that it be done.

paulp committed:
Removed code from scaladoc.

Nobody is immune!

paulp committed:
Remove code from misc bits of the compiler.

      They are everywhere.
    They defy categorization.
      They are... M I S C

paulp committed:
Remove code from compiler central.

All those old-timey methods whose melodies have become
unfashionable.

paulp commented:
I moved the Settings to its own pull request, and subdivided the removals into about ten commits.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/929/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/929/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1634/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1634/

paulp commented:
"Captain Spurious! Have mercy on us!"

paulp committed:
Merge remote-tracking branch 'paulp/mutable-settings' into unused-bonanza

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1635/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1635/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/930/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/930/

adriaanm commented:
We must save ourselves from the clutches of Lt. Bitrot. It's a question of life v death.
I know this represents a lot of work, and I do not wish to have the anti-bitrot bitrot. 

However.

I will not merge this until representatives from all our core projects (SBT, IDE, Slick,...) have signed off.
Those affected by a big change like this need to be included. It's everybody's problem -- both ways.
- SBT should be represented by a test case. But will all of it build? /CC @harrah
- The IDE is definitely going to be affected. /CC @dragos 
- The Slick project might want to check this out as well. /CC @szeiger

Let's make this a coordinated effort.

It would be best if we could have a ro-bot answer these questions for us. /CC @jsuereth 
In the mean time the humans shall have come out of hiding.

harrah commented:
I haven't gotten all of sbt compiling against master yet, but it doesn't look like these changes are the cause of any compilation issues.  The parts that interface with the compiler, including the compiler interface, have compiled successfully.

adriaanm commented:
Cool, thanks Mark.

jsuereth commented:
Yes, SBT's interface compiles, the beautiful benefit of having tat in the ANT build.

As far as the IDE / Slick, I can attempt to push slick into the community build and see what happens.  Note:  My talk for Devoxx is today, so most likely I won't get around to it until tomorrow.

For the IDE, we need someone on @dragos team to run the build, and the tests.

jboner commented:
Awesome job Paul.

szeiger commented:
After spending almost the whole day on getting Slick to compile against a locally built Scala version again (with a multi-project, multi-configuration sbt build that needs to pull in scala-compiler.jar for the macros stuff), I still have no conclusive result. Compilation of Slick against this PR fails in the same way as it does against the previous commit 91abce5d1a5c5681dbac565cb4cbcf81ff54bb60, with some completely nonsensical (at least to me) errors: http://pastie.org/5377836

paulp commented:
I reached the same conclusion with slick, although for once it sounds like I was able to do something with sbt faster than someone else. I'm looking at the error.

dragos commented:
I will have a look at this tomorrow morning. Would it be possible to push pull-request artifacts to a maven repository that's accessible over the intertubes? Then I can prepare a pull-request validator IDE job on our Jenkins that you can run. Or we can do it on the EPFL one. The only thing I need is a repo that I can point our build to, and where the scala compiler can be found.

paulp commented:
WRT debugging slick: I don't suppose anyone knows how to make sbt use the freaking scalaHome for scala-reflect.jar? It happily mixes library and compiler jars from scalaHome with scala-reflect.jar RC2. This places a hard limit on what I can examine before I start exploding with NoSuchMethod.

paulp commented:
@dragos there's a jenkins plugin for exposing the build artifacts as a maven repo.

szeiger commented:
@paulp It should work nicely now. Just switch from publishedScalaSettings to localScalaSettings in https://github.com/slick/slick/commit/f3462096418d9b9cba6573cf1f2dee1c084c6446

paulp commented:
https://github.com/slick/slick/pull/63

paulp commented:
Slick regression was in 97ede5a64a, I'm looking at it.  [...] #1625

adriaanm commented:
gentle ping @dragos -- just trying to keep the train moving

dragos commented:
@adriaanm, sorry, I've been on the go for the past 3 days with limited connectivity. I'll follow up.

dotta commented:
Here goes the report for building the IDE against this checkout:

## Compiling dependencies
- Success for scalariform, sbinary and sbt (not the whole sbt, but the artifacts needed for incremental compilation).
- Failure for scala-refactoring.

Here are the errors reported while compiling scala-refactoring:

> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-refactoring/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/analysis/NameValidation.scala:34: error: value isIdentifier is not a member of object scala.tools.nsc.ast.parser.Tokens
> [ERROR]       val firstTokenIsIdentifier = Tokens.isIdentifier(scanner.token)
> [ERROR]                                           ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-refactoring/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/sourcegen/PrettyPrinter.scala:209: error: value isOperatorName is not a member of PrettyPrinter.this.global.Name
> [ERROR]       def isOperator(n: Name) = n.isOperatorName && n != nme.CONSTRUCTOR
> [ERROR]                                   ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-refactoring/org.scala-refactoring.library/src/main/scala/scala/tools/refactoring/sourcegen/ReusingPrinter.scala:966: error: value isOperatorName is not a member of ReusingPrinter.this.global.Name
> [ERROR]         case Apply(Select(_, name), _) if rhs.pos.isTransparent && name.isOperatorName =>
> [ERROR]                                                                         ^
> [ERROR] three errors found

Maybe Mirko (@misto) can help with those.

## Compiling the IDE

To start the IDE's build, I've used a scala-refactoring artifact compiled against a different Scala checkout, but that should not matter if what we are after are compilation errors. Here are the reported errors:

> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:53: error: value toChars is not a member of ScalaIndexBuilder.this.Name
> [ERROR]       case Ident(id)                           => id.toChars
> [ERROR]                                                      ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:54: error: value toChars is not a member of ScalaIndexBuilder.this.Name
> [ERROR]       case Select(_, name)                     => name.toChars
> [ERROR]                                                        ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:55: error: value toChars is not a member of ScalaIndexBuilder.this.Name
> [ERROR]       case AppliedTypeTree(fun: RefTree, args) => fun.name.toChars
> [ERROR]                                                            ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:91: error: value toChars is not a member of ScalaIndexBuilder.this.TypeName
> [ERROR]         c.name.toChars,
> [ERROR]                ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:106: error: value toChars is not a member of ScalaIndexBuilder.this.Name with m.name.ThisNameType
> [ERROR]         m.name.append("$").toChars,
> [ERROR]                            ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:117: error: value toChars is not a member of ScalaIndexBuilder.this.TermName
> [ERROR]         m.name.toChars,
> [ERROR]                ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:128: error: value toChars is not a member of ScalaIndexBuilder.this.TermName
> [ERROR]         nme.getterName(v.name).toChars,
> [ERROR]                                ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:136: error: value toChars is not a member of ScalaIndexBuilder.this.TermName
> [ERROR]           nme.getterToSetter(nme.getterName(v.name)).toChars,
> [ERROR]                                                      ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:145: error: value toChars is not a member of ScalaIndexBuilder.this.Name
> [ERROR]       val name = if(nme.isConstructorName(d.name)) enclClassNames.head else d.name.toChars
> [ERROR]                                                                                    ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:164: error: value toChars is not a member of ScalaIndexBuilder.this.TypeName
> [ERROR]         td.name.toChars,
> [ERROR]                 ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:194: error: value toChars is not a member of ScalaIndexBuilder.this.Name
> [ERROR]           indexer.addMethodReference(tree.name.toChars, i)
> [ERROR]                                                ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:209: error: value toChars is not a member of ScalaIndexBuilder.this.TypeName
> [ERROR]         case cd : ClassDef => inClass(cd.name.toChars) { super.traverse(tree) }
> [ERROR]                                               ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:210: error: value toChars is not a member of ScalaIndexBuilder.this.Name with md.name.ThisNameType
> [ERROR]         case md : ModuleDef => inClass(md.name.append("$").toChars) { super.traverse(tree) }
> [ERROR]                                                            ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:228: error: value toChars is not a member of ScalaIndexBuilder.this.Name
> [ERROR]           val name = rt.name.toChars
> [ERROR]                              ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaIndexBuilder.scala:231: error: value toChars is not a member of ScalaIndexBuilder.this.TermName
> [ERROR]           if(nme.isSetterName(rt.name)) indexer.addFieldReference(nme.setterToGetter(rt.name).toChars)
> [ERROR]                                                                                               ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaMatchLocator.scala:204: error: value toChars is not a member of sym.NameType
> [ERROR]       if (!pat.matchesName(pat.selector, sym.name.toChars) || !sym.pos.isDefined) {
> [ERROR]                                                   ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaMatchLocator.scala:255: error: value toChars is not a member of ScalaMatchLocator.this.Name
> [ERROR]       lazy val nameNoMatch = !pat.matchesName(searchedVar, s.name.toChars)
> [ERROR]                                                                   ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaMatchLocator.scala:256: error: value toChars is not a member of ScalaMatchLocator.this.Name
> [ERROR]       lazy val varNoMatch = !pat.matchesName(CharOp.concat(searchedVar, "_$eq".toCharArray), s.name.toChars)
> [ERROR]                                                                                                     ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaSelectionEngine.scala:118: error: value toChars is not a member of compiler.Symbol#NameType with compiler.Name
> [ERROR]           name.toChars,
> [ERROR]                ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaStructureBuilder.scala:181: error: value toChars is not a member of ScalaStructureBuilder.this.TermName
> [ERROR]           val paramNames = Array(fps.map(n => nme.getterName(n.name).toChars) : _*)
> [ERROR]                                                                      ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaStructureBuilder.scala:430: error: value toChars is not a member of ScalaStructureBuilder.this.Symbol#NameType
> [ERROR]               tpElementInfo.bounds = parents.map(_.typeSymbol.name.toChars).toArray
> [ERROR]                                                                    ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/javaelements/ScalaStructureBuilder.scala:679: error: value toChars is not a member of ScalaStructureBuilder.this.TermName
> [ERROR]           val originalParamNames = fps.map(n => nme.getterName(n.name).toChars)
> [ERROR]                                                                        ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/markoccurrences/ScalaOccurrencesFinder.scala:66: error: value isOperatorName is not a member of x$3.NameType
> [ERROR]           Option(selectedTree.symbol) filter (!_.name.isOperatorName) map { sym =>
> [ERROR]                                                       ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/properties/IDESettings.scala:23: error: not found: value Xchecknull
> [ERROR]       List(checkInit, Xchecknull, elidebelow,
> [ERROR]                         ^
> [ERROR] /Users/mirco/Projects/ide/uber-build/scala-ide/org.scala-ide.sdt.core/src/scala/tools/eclipse/properties/IDESettings.scala:30: error: not found: value Ywarndeadcode
> [ERROR] 25 errors found

It is probably worth pointing out that _the same_ Scala IDE codebase is currently used to build against Scala 2.9, 2.10 and 2.11. If it helps, I think it would be ok from our side to create a new branch for the next next Scala IDE release, which would only need to be compiled against both 2.10 and 2.11.

paulp commented:
Thanks a lot mirco. We can worry about the long-term handling of these things later; for now I will mark them, put them all back, and open a rebased pull request.

paulp commented:
Tenting for remodeling. Will reopen under new ownership.

dotta commented:
> Thanks a lot mirco. We can worry about the long-term handling of these things later; for now I will mark them, put them all back, and open a rebased pull request.

Sounds good to me. Please, add deprecation warnings where it makes sense (I'm guessing that there must be an alternative to `Name.toChars`).

adriaanm commented:
awesome -- looking forward to the grand reopening

paulp commented:
Passed our health inspection; now taking orders at our new address, #1648. No shirt, no shoes, we don't care.
